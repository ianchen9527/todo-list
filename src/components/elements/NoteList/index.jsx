import React from "react";
import styles from "./index.module.scss";
import PropTypes from "prop-types";
import cx from "classnames";

const getNoteWrapperClassName = (disabled) =>
  cx({
    [styles["note-wrapper"]]: true,
    [styles["disabled"]]: disabled,
  });

const renderLoadingView = () => <div className={styles["loading-view"]} />;

const renderList = ({ notes, disabled, onNoteClick }) => {
  const noteWrapperClassName = getNoteWrapperClassName(disabled);

  return notes.length > 0 ? (
    notes.map((note) => (
      <div
        className={noteWrapperClassName}
        onClick={() => onNoteClick(note.id)}
        key={note.id}
      >
        {note.title}
      </div>
    ))
  ) : (
    <div className={styles["place-holder"]}>
      Please click the "New Note" button to add a new note...
    </div>
  );
};

const NoteList = (props) => (
  <div className={styles["wrapper"]}>
    <div className={styles["title"]}>Note List</div>
    {props.loading ? renderLoadingView() : renderList(props)}
  </div>
);

NoteList.oropTypes = {
  notes: PropTypes.arrayOf(
    PropTypes.shape({ title: PropTypes.string, id: PropTypes.number })
  ),
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  onNoteClick: PropTypes.func,
};

NoteList.defaultProps = {
  notes: [],
  disabled: false,
  loading: false,
  onNoteClick: null,
};

export default NoteList;
