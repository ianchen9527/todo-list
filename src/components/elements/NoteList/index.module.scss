@import "../../../stylesheets/constants/colors.scss";

.wrapper {
  width: 100%;
  height: 100%;
  overflow: auto;
  box-sizing: border-box;
  padding: 10px 30px;

  .title {
    color: $black;
    font-family: "Roboto";
    font-weight: bold;
    font-size: 20px;
    line-height: 30px;
    margin-bottom: 15px;
  }

  .loading-view {
    width: 60px;
    height: 60px;
    display: block;
    background-size: 60px 60px;
    background-image: url("../../../assets/images/loading.gif");
  }

  .place-holder {
    color: $black;
    font-family: "Roboto";
    font-weight: bold;
    font-size: 16px;
    line-height: 24px;
  }

  .note-wrapper {
    border: 1px solid $black;
    padding: 15px;
    border-radius: 5px;
    box-sizing: border-box;
    margin-bottom: 15px;
    background-color: $button;
    cursor: pointer;
    color: $black;
    font-family: "Roboto";
    font-size: 14px;
    line-height: 21px;

    &:hover {
      background-color: $button-hover;
    }

    &.disabled {
      opacity: 0.5;
      cursor: default;

      &:hover {
        background-color: $button;
      }
    }
  }
}
