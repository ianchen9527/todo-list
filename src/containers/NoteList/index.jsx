import { compose } from "redux";
import { connect } from "react-redux";
import { getNotes, getOnNoteClick } from "../../selectors/containers/noteList";
import NoteList from "../../components/elements/NoteList";
import withLifecycle from "@hocs/with-lifecycle";
import { getOnDidMount } from "../../sagas/noteList";

const mapStateToProps = (state) => ({
  notes: getNotes(state),
});

const mapDispatchToProps = (dispatch) => ({
  onNoteClick: getOnNoteClick(dispatch),
  dispatch,
});

const lifeCycle = {
  onDidMount: getOnDidMount(),
};

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withLifecycle(lifeCycle),
)(NoteList);
