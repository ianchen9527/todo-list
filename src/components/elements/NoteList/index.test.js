import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import NoteList from "./index";

describe("NoteList", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<NoteList />);
  });

  describe("default props", () => {
    it("should render without crashing", () => {
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe("props.loading", () => {
    let loading;
    describe("is true", () => {
      beforeEach(() => {
        loading = true;
        wrapper.setProps({ loading });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });
    });

    describe("is false", () => {
      beforeEach(() => {
        loading = false;
        wrapper.setProps({ loading });
      });

      it("should render without crashing", () => {
        expect(toJSON(wrapper)).toMatchSnapshot();
      });

      describe("props.notes", () => {
        let notes;
        describe("is an empty array", () => {
          beforeEach(() => {
            notes = [];
            wrapper.setProps({ notes });
          });

          it("should render without crashing", () => {
            expect(toJSON(wrapper)).toMatchSnapshot();
          });
        });

        describe("is not an empty array", () => {
          beforeEach(() => {
            notes = [
              { id: 1, title: "title 1" },
              { id: 2, title: "title 2" },
              { id: 3, title: "title 3" },
            ];
            wrapper.setProps({ notes });
          });

          it("should render without crashing", () => {
            expect(toJSON(wrapper)).toMatchSnapshot();
          });

          describe("props.onNoteClick", () => {
            let onNoteClick, random;
            beforeEach(() => {
              onNoteClick = jest.fn();
              random = Math.floor(Math.random() * notes.length);
              wrapper.setProps({ onNoteClick });
            });

            afterEach(() => {
              onNoteClick.mockClear();
            });

            describe("when any of .note-wrapper is changed", () => {
              beforeEach(() => {
                wrapper.find(".note-wrapper").at(random).simulate("click");
              });

              it("should be triggered with correspond id", () => {
                expect(onNoteClick).toHaveBeenCalledWith(notes[random].id);
              });
            });
          });
        });
      });
    });
  });
});
