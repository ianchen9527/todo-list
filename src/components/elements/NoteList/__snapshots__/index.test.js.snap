// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteList default props should render without crashing 1`] = `
<div
  className="wrapper"
>
  <div
    className="title"
  >
    Note List
  </div>
  <div
    className="place-holder"
  >
    Please click the "New Note" button to add a new note...
  </div>
</div>
`;

exports[`NoteList props.loading is false props.notes is an empty array should render without crashing 1`] = `
<div
  className="wrapper"
>
  <div
    className="title"
  >
    Note List
  </div>
  <div
    className="place-holder"
  >
    Please click the "New Note" button to add a new note...
  </div>
</div>
`;

exports[`NoteList props.loading is false props.notes is not an empty array should render without crashing 1`] = `
<div
  className="wrapper"
>
  <div
    className="title"
  >
    Note List
  </div>
  <div
    className="note-wrapper"
    key="1"
    onClick={[Function]}
  >
    title 1
  </div>
  <div
    className="note-wrapper"
    key="2"
    onClick={[Function]}
  >
    title 2
  </div>
  <div
    className="note-wrapper"
    key="3"
    onClick={[Function]}
  >
    title 3
  </div>
</div>
`;

exports[`NoteList props.loading is false should render without crashing 1`] = `
<div
  className="wrapper"
>
  <div
    className="title"
  >
    Note List
  </div>
  <div
    className="place-holder"
  >
    Please click the "New Note" button to add a new note...
  </div>
</div>
`;

exports[`NoteList props.loading is true should render without crashing 1`] = `
<div
  className="wrapper"
>
  <div
    className="title"
  >
    Note List
  </div>
  <div
    className="loading-view"
  />
</div>
`;
